clippy_lints/src/lifetimes_bound_nested_ref.rs

```diff
@@ -0,0 +1,222 @@
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// pub const fn lifetime_translator<'a, 'b, T>(val_a: &'a &'b (), val_b: &'b T) -> &'a T {...}
+    /// ```
+
+    #[clippy::version = "1.78.0"]
+    pub REMOVE_REDUNDANT_LIFETIMES_BOUND_NESTED_REF_FN,
+    nursery,
+    "suggest to remove generic lifetime bounds implied by nested references in function arguments and return value"
+}
+
+pub struct LifetimesBoundNestedRef;
+
+impl_lint_pass!(LifetimesBoundNestedRef => [
+    ADD_REDUNDANT_LIFETIMES_BOUND_NESTED_REF_FN,
+    REMOVE_REDUNDANT_LIFETIMES_BOUND_NESTED_REF_FN,
+]);
+
+#[derive(Debug)]
+struct BoundLifetimePair<'a> {
+    long_lifetime: &'a Lifetime,
+    outlived_lifetime: &'a Lifetime,
+}
+
+impl<'a> PartialEq for BoundLifetimePair<'a> {
+    fn eq(&self, other: &Self) -> bool {
+        self.long_lifetime.res.eq(&other.long_lifetime.res)
+            && self.outlived_lifetime.res.eq(&other.outlived_lifetime.res)
+    }
+
+    fn ne(&self, other: &Self) -> bool {
+        !self.eq(other)
+    }
+}
+
+impl<'a> BoundLifetimePair<'a> {
+    fn as_bound_declaration(&'a self) -> String {
+        format!(
+            "{}: {}",
+            self.long_lifetime.ident.name, self.outlived_lifetime.ident.name,
+        )
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for LifetimesBoundNestedRef {
+    fn check_fn<'tcx2>(
+        &mut self,
+        ctx: &LateContext<'tcx2>,
+        fn_kind: rustc_hir::intravisit::FnKind<'tcx2>,
+        fn_decl: &'tcx2 rustc_hir::FnDecl<'tcx2>,
+        _body: &'tcx2 rustc_hir::Body<'tcx2>,
+        _span: rustc_span::Span,
+        _local_def_id: rustc_span::def_id::LocalDefId,
+    ) {
+        let FnKind::ItemFn(_ident, generics, _fn_header) = fn_kind else {
+            return;
+        };
+        if generics.predicates.is_empty() && fn_decl.inputs.is_empty() {
+            return;
+        }
+        // collect declared predicate bounds on lifetime pairs
+        let mut declared_bounds = Vec::<BoundLifetimePair<'_>>::new();
+        for where_predicate in generics.predicates {
+            declared_bounds.extend(get_declared_bounds(*where_predicate));
+        }
+        // collect bounds implied by nested references with lifetimes in arguments
+        let mut implied_bounds = Vec::<BoundLifetimePair<'_>>::new();
+        for input_ty in fn_decl.inputs {
+            implied_bounds.extend(get_nested_ref_implied_bounds(input_ty));
+        }
+        // and for function return type
+        if let rustc_hir::FnRetTy::Return(ret_ty) = fn_decl.output {
+            implied_bounds.extend(get_nested_ref_implied_bounds(ret_ty));
+        }
+
+        let implied_bounds = implied_bounds;
+
+        for implied_bound in &implied_bounds {
+            if !declared_bounds.contains(implied_bound) {
+                span_lint(
+                    ctx,
+                    ADD_REDUNDANT_LIFETIMES_BOUND_NESTED_REF_FN,
+                    generics.span,
+                    &format!(
```

Comment:
The one thing I think should happen before landing this is to make it a structured suggestion, so people can just rustfix their crates instead of having to manually adjust possibly hundreds of functions